---
title: "p2954 Analysis Report"
author: "Witold Wolski"
date: "1 February 2019"
output:
  pdf_document: default
  html_document: default
bibliography: bibliography.bib
---


File `MQSummary.pdf` summaries information about the dataset including raw file to sample name mappings, numbers of identified proteins, intensity distribution of peptides within a condition and samples as well as coefficient of variation (CV) and standard deviation of peptide measurements. 


File `SummarizeFiltering.pdf` contains information about the number of quantifiable peptides and proteins used for the statistical analysis. Furthermore, peptides were filtered. A peptide is considered quantified if it was observed in more than $50%$ of samples in one of the conditions. We required $2$ quantified peptides per protein.  Proteins with only one quantified peptide were removed.
This document also summarises intensity normalization. Normalization is applied to the peptide intensities to remove systematic differences in protein abundance due to different sample concentrations, or different amount of sample loaded on column. Normalization is important, so that true differentially expressed proteins can be detected. To do this the z-score of the log2 transformed intensities is computed, which is updated by the average of the standard deviation of the log2 transformed intensities in all samples. After normalization all samples have a similar distribution. 


# Modelling peptide intensities using mixed linear models to obtain protein fold change estimates between condition.

Using Linear Mixed Effect Models [@BatesBokler2015] we study how the factor IVD treatment can explain peptide intensities. 
We report fold-change estimate per protein. The peptide measurements are treated as repeated measurements of the same protein and are modeled using a random effect.
Since we do have repeated measurements on the same patients, we model the correlation of these within patient measurement using an additional random effect. We tested the significance of this random effect using the likelihood ratio test.



## Model 1 - fixed effect Condition - IVD treatment, random effect peptide

$$
y_{ijkl} = \alpha_i + a_l  + \epsilon_{il}
$$

with:

$$\alpha_i$$ modelling the conditions. $$a_l \propto N(0, s_a), b_k \propto N(0, s_a), \epsilon \propto N(0, s_e) $$
The the peptides $a_l$ represent repeated measurements and, $l$ iterates over all peptides of the protein.

The files:

- `Contrasts_Auto_SignificanceValues_f_Condition_r_peptide.csv` 
- `Contrasts_SignificanceValues_f_Condition_r_peptid_PIVOT.csv` 

contain fold changes and p-values when comparing the conditions based on this model.

The file `Group_Averages_f_Condition_r_peptide.csv` contains the protein intensities for each condition estimated using this model. 

The files 

- `Contrasts_Auto_histogram_f_Condition_r_peptide.pdf` 
- `Contrasts_Auto_Volcano_f_Condition_r_peptide.pdf` 

contain visualization of fold-change estimates and p.values.


## Model 2 - fixed effect Condition - IVD treatment, random effect peptide and patient. 

$$
y_{ijkl} = \alpha_i + a_l + b_k + \epsilon_{il}
$$
with  $b_k \propto N(0, s_a)$. Differences among patients are modeled using the coefficient $b_k$ where $k$ iterates over all patients.


The files 

- `Contrasts_Auto_SignificanceValues_f_Condition_r_peptide_r_patient.csv`
- `Contrasts_SignificanceValues_f_Condition_r_peptid_r_patient_PIVOT.csv` 

contain fold changes and p-values when comparing the conditions based on this model.

The file `Group_Averages_f_Condition_r_peptide_r_patient.csv` contains the protein intensities for each condition estimated using this model. 

The files 

- `Contrasts_Auto_histogram_f_Condition_r_peptide_r_patient.pdf`
- `Contrasts_Auto_Volcano_f_Condition_r_peptide_r_patient.pdf` 

contain visualization of fold-change estimates and p.values.


# Likelihood ratio test.


All `.csv` files contain a column `likelihood_ratio_test.pValue`. This `p.value` tests if Model 2 significantly better models the data than Model 1. If it does not than Model 1 should be preferred (parsimony principle).

Since random effect was highly significant for most proteins, we decided to use the results of this model (file : `Contrasts_SignificanceValues_f_Condition_r_peptide_patient_PIVOT.csv`) 


# Data Interpretation

For interpreting the results, the protein IDs can be either sorted by the fold change estimates or `p.value`. Large positive or negative fold changes typically result in small p-values therefore we suggest sorting by foldchange.

The IDs sorted by fold change can then be subjected to gene set enrichment analysis (GSEA). Alternatively, a subset filtered by q-value can be analysed using over-representation analysis (ORA). The web application WebGestalt (WEB-based GEne SeT AnaLysis Toolkit) <http://www.webgestalt.org> implements both of these methods (and even more) for the most popular organisms.

Overrepresentation analysis is performed on biological functional categories (e.g., biological processes of gene ontology annotations) or on biological pathways (e.g., KEGG or Wikipathways). Using such methods allows identifying functions or pathways for proteins in the submitted list. For the correct usage and interpretation of the results from such an analysis, it is essential to specify the background proteome. The background proteome is the list of all proteins identified in your experiment.


A further resource to analyze the results is the STRING database <https://string-db.org>. It reports known and predicted interactions for proteins in the submitted list.


The FGCZ can support you, with the interpretation of your quantitative proteomics data or with a more customized analysis. Further visualization of the data, targeted to your audience, e.g., receiver operator curves (ROC) or MA-plots, can be generated. You can reach the proteome-bioinformatics team at <protinf@fgcz.uzh.ch>.


# Disclaimer


The obtained results should be validated orthogonal as well (e.g. with Western blots). The Functional Genomics Center Zurich does not provide any kind of guarantee for the validity of the results.


For questions and improvement suggestions, with respect to this report, please do contact <protinf@fgcz.uzh.ch>.


# Session Inforamtion


```{r echo=FALSE}
library(LFQService)
pander::pander(sessionInfo())
```


